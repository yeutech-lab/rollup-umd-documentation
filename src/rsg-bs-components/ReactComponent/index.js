export { default } from 'rsg-components/ReactComponent/ReactComponentRenderer';
