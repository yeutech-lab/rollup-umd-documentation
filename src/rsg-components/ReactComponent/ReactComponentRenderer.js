import React from 'react';
import PropTypes from 'prop-types';
import Header from 'bootstrap-styled/lib/Header';
import omit from 'lodash.omit';
import styled from 'styled-components';
import mapToCssModules from 'map-to-css-modules/lib';
import cn from 'classnames';
import Pathline from '../Pathline';

export const defaultProps = {
  theme: {
    styleguide: {
      '$rsg-react-component-margin': '0 0 40px 0',
      '$rsg-react-component-header-margin': '0 0 16px 0',
      '$rsg-react-component-docs-color': '#333',
      '$rsg-react-component-docs-font-size': '16px',
      '$rsg-react-component-tabs-margin': '0 0 16px 0',
      '$rsg-react-component-tabs-button-margin': '0 0 8px 0',
    },
  },
};
/* eslint-disable react/require-default-props */
export const propTypes = {
  /**
   * @ignore
   */
  className: PropTypes.string, // eslint-disable-line react/require-default-props
  /** Name used in id. */
  name: PropTypes.string.isRequired,
  /** Heading element to be rendered. */
  heading: PropTypes.node.isRequired,
  /** Component directory file path. */
  filepath: PropTypes.string,
  /** Component directory path line. */
  pathLine: PropTypes.string,
  /** Tab buttons elements to be rendered. */
  tabButtons: PropTypes.node,
  /** Tab body element to be rendered. */
  tabBody: PropTypes.node,
  /** Description element to be rendered. */
  description: PropTypes.node,
  /** Documentation element to be rendered. */
  docs: PropTypes.node,
  /** Examples elements to be rendered. */
  examples: PropTypes.node,
  /** Toggle isolated style. */
  isolated: PropTypes.bool,
  /** Theme variables. Can be: */
  theme: PropTypes.shape({
    styleguide: PropTypes.shape({
      '$rsg-react-component-margin': PropTypes.string,
      '$rsg-react-component-header-margin': PropTypes.string,
      '$rsg-react-component-docs-color': PropTypes.string,
      '$rsg-react-component-docs-font-size': PropTypes.string,
      '$rsg-react-component-tabs-margin': PropTypes.string,
      '$rsg-react-component-tabs-button-margin': PropTypes.string,
    }),
  }),
  /**
   * Replace or remove a className from the component.
   * See example <a href="https://www.npmjs.com/package/map-to-css-modules" target="_blank">here</a>.
   */
  cssModule: PropTypes.object, // eslint-disable-line react/require-default-props
};
/* eslint-enable react/require-default-props */


const ReactComponentRendererUnstyled = (props) => {
  const {
    className,
    name,
    heading,
    pathLine,
    description,
    docs,
    examples,
    tabButtons,
    tabBody,
    cssModule,
    ...attributes
  } = omit(props, ['theme']);

  return (
    <div
      id={`${name}-container`}
      className={mapToCssModules(cn(className, 'rsg-react-component'), cssModule)}
      {...attributes}
    >
      <Header className="rsg-react-component-header">
        {heading}
        {pathLine && <Pathline>{pathLine}</Pathline>}
      </Header>
      {(description || docs) && (
        <div className="rsg-react-component-docs">
          {description}
          {docs}
        </div>
      )}
      {tabButtons && (
        <div className="rsg-react-component-tabs">
          <div className="rsg-react-component-tabs-button">{tabButtons}</div>
          {tabBody}
        </div>
      )}
      {examples}
    </div>
  );
};

ReactComponentRendererUnstyled.defaultProps = defaultProps;
ReactComponentRendererUnstyled.propTypes = propTypes;

const ReactComponentRenderer = styled(ReactComponentRendererUnstyled)` 
  ${(props) => `
    &.rsg-react-component {
      margin: ${props.theme.styleguide['$rsg-react-component-margin']};
      .rsg-react-component-header {
        margin: ${props.theme.styleguide['$rsg-react-component-header-margin']};
      }
      .rsg-react-component-docs {
        color: ${props.theme.styleguide['$rsg-react-component-docs-color']};
        font-size: ${props.theme.styleguide['$rsg-react-component-docs-font-size']};
      }
      .rsg-react-component-tabs {
        margin: ${props.theme.styleguide['$rsg-react-component-tabs-margin']};
        .rsg-react-component-tab-buttons {
          margin: ${props.theme.styleguide['$rsg-react-component-tabs-button-margin']};
        }
      }
    }
 `}
`;

ReactComponentRenderer.defaultProps = defaultProps;
ReactComponentRenderer.propTypes = propTypes;

export default ReactComponentRenderer;
