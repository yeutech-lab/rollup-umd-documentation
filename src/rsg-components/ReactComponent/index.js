export { default } from 'rsg-components/ReactComponent/ReactComponent';
