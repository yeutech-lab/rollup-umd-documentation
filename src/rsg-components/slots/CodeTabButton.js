import React from 'react';
import PropTypes from 'prop-types';
import TabButton from 'rsg-components/TabButton';

const CodeTabButton = props => <TabButton {...props}>View Code</TabButton>;

CodeTabButton.propTypes = {
	onClick: PropTypes.func.isRequired,
	name: PropTypes.string.isRequired,
	active: PropTypes.bool,
};

export default CodeTabButton;
