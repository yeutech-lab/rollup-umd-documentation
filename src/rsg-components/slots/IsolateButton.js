import React from 'react';
import PropTypes from 'prop-types';
import MdFullscreen from 'react-icons/lib/md/fullscreen';
import MdFullscreenExit from 'react-icons/lib/md/fullscreen-exit';
import ToolbarButton from 'rsg-components/ToolbarButton';
import getUrl from '../../utils/getUrl';

const IsolateButton = ({ name, example, isolated }) =>
	isolated ? (
		<ToolbarButton href={getUrl({ anchor: true, slug: '/' })} title="Show all components">
			<MdFullscreenExit />
		</ToolbarButton>
	) : (
		<ToolbarButton href={getUrl({ name, example, isolated: true })} title="Open isolated">
			<MdFullscreen />
		</ToolbarButton>
	);

IsolateButton.propTypes = {
	name: PropTypes.string.isRequired,
	example: PropTypes.number,
	isolated: PropTypes.bool,
};

export default IsolateButton;
