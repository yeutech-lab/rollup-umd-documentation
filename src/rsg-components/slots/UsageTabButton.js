import React from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';
import TabButton from '../TabButton'; // eslint-disable-line no-unused-vars

const UsageTabButton = (props) => {
  const component = props.props;
  const showButton = component.props || (component.methods && component.methods.length > 0);
  return showButton ? <TabButton {...props}>Props & methods</TabButton> : null;
};

UsageTabButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  props: PropTypes.shape({
    props: PropTypes.array,
    methods: PropTypes.array,
  }).isRequired,
  active: PropTypes.bool,
};

export default UsageTabButton;
