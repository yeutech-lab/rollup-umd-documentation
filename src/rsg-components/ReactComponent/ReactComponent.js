import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Examples from 'rsg-components/Examples';
import SectionHeading from 'rsg-components/SectionHeading';
import JsDoc from 'rsg-components/JsDoc';
import Markdown from 'rsg-components/Markdown';
import Slot from 'rsg-components/Slot';
import ReactComponentRenderer from 'rsg-components/ReactComponent/ReactComponentRenderer';
import { DOCS_TAB_USAGE } from '../slots';
import { DisplayModes } from '../../consts';

const ExamplePlaceholder =
	process.env.STYLEGUIDIST_ENV !== 'production'
		? require('rsg-components/ExamplePlaceholder').default
		: () => <div />;

export default class ReactComponent extends Component {
	static propTypes = {
		component: PropTypes.object.isRequired,
		depth: PropTypes.number.isRequired,
	};
	static contextTypes = {
		config: PropTypes.object.isRequired,
		displayMode: PropTypes.string,
	};

	constructor(props, context) {
		super(props, context);
		const { showUsage } = context.config;

		this.handleTabChange = this.handleTabChange.bind(this);

		this.state = {
			activeTab: showUsage ? DOCS_TAB_USAGE : undefined,
		};
	}

	handleTabChange(name) {
		this.setState(state => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	}

	render() {
		const { activeTab } = this.state;
		const { displayMode } = this.context;
		const { component, depth } = this.props;
		const { name, slug, filepath, pathLine } = component;
		const { description, examples = [], tags = {} } = component.props;
		if (!name) {
			return null;
		}

		return (
			<ReactComponentRenderer
				name={name}
				slug={slug}
				filepath={filepath}
				pathLine={pathLine}
				docs={<JsDoc {...tags} />}
				description={description && <Markdown text={description} />}
				heading={
					<SectionHeading
						id={slug}
						deprecated={!!tags.deprecated}
						slotName="componentToolbar"
						slotProps={{
							...component,
							isolated: displayMode !== DisplayModes.all,
						}}
						depth={depth}
					>
						{name}
					</SectionHeading>
				}
				examples={
					examples.length > 0 ? (
						<Examples examples={examples} name={name} />
					) : (
						<ExamplePlaceholder name={name} />
					)
				}
				tabButtons={
					<Slot
						name="docsTabButtons"
						active={activeTab}
						props={{ ...component, onClick: this.handleTabChange }}
					/>
				}
				tabBody={<Slot name="docsTabs" active={activeTab} onlyActive props={component} />}
			/>
		);
	}
}
