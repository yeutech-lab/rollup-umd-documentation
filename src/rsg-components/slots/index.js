import Editor from 'rsg-components/Editor';
import Usage from 'rsg-components/Usage';
import IsolateButton from './IsolateButton';
import CodeTabButton from './CodeTabButton';
import UsageTabButton from './UsageTabButton';

export const EXAMPLE_TAB_CODE_EDITOR = 'rsg-code-editor';
export const DOCS_TAB_USAGE = 'rsg-usage';

export default ({ pagePerSection }) => {
	const toolbar = pagePerSection ? [] : [IsolateButton];

	return {
		sectionToolbar: toolbar,
		componentToolbar: toolbar,
		exampleToolbar: toolbar,
		exampleTabButtons: [
			{
				id: EXAMPLE_TAB_CODE_EDITOR,
				render: CodeTabButton,
			},
		],
		exampleTabs: [
			{
				id: EXAMPLE_TAB_CODE_EDITOR,
				render: Editor,
			},
		],
		docsTabButtons: [
			{
				id: DOCS_TAB_USAGE,
				render: UsageTabButton,
			},
		],
		docsTabs: [
			{
				id: DOCS_TAB_USAGE,
				render: Usage,
			},
		],
	};
};
