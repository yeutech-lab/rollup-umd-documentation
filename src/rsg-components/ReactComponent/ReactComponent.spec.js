import React from 'react';
import slots, { DOCS_TAB_USAGE } from '../slots';
import ReactComponent from './ReactComponent';
import { ReactComponentRenderer } from './ReactComponentRenderer';
import { DisplayModes } from '../../consts';

const options = {
	context: {
		config: {
			showUsage: false,
		},
		displayMode: DisplayModes.all,
		slots,
	},
	metadata: {},
};

const component = {
	name: 'Foo',
	slug: 'foo',
	pathLine: 'foo/bar.js',
	props: {
		description: 'Bar',
		methods: [],
		examples: [],
	},
	metadata: {},
};
const componentWithEverything = {
	name: 'Foo',
	slug: 'foo',
	pathLine: 'foo/bar.js',
	props: {
		description: 'Bar',
		methods: [
			{
				name: 'set',
				params: [
					{
						name: 'newValue',
						description: 'New value for the counter.',
						type: { name: 'Number' },
					},
				],
				returns: null,
				description: 'Sets the counter to a particular value.',
			},
		],
		examples: [
			{
				type: 'code',
				content: '<button>OK</button>',
				evalInContext: () => {},
			},
			{
				type: 'markdown',
				content: 'Hello *world*!',
			},
		],
	},
	metadata: {
		tags: ['one', 'two'],
	},
};

describe('ReactComponent', () => {
	it('should render an example placeholder', () => {
		const actual = shallow(<ReactComponent component={component} depth={3} />, options);

		const props = actual.prop('examples').props;
		expect(props.name).toBeTruthy();
		expect(props.examples).toBeFalsy();
	});

	it('should render examples', () => {
		const actual = shallow(
			<ReactComponent component={componentWithEverything} depth={3} />,
			options
		);

		const props = actual.prop('examples').props;
		expect(props.name).toBeTruthy();
		expect(props.examples).toBeTruthy();
	});

	it('should pass rendered description, usage, examples, etc. to the renderer', () => {
		const actual = shallow(
			<ReactComponent component={componentWithEverything} depth={3} />,
			options
		);

		expect(actual).toMatchSnapshot();
	});

	it('should render usage closed by default when showUsage config options is false', () => {
		const actual = shallow(
			<ReactComponent component={componentWithEverything} depth={3} />,
			options
		);

		expect(actual.prop('tabButtons').props.active).toBeFalsy();
		expect(actual.prop('tabBody').props.active).toBeFalsy();
	});

	it('should render usage opened by default when showUsage config options is true', () => {
		const actual = shallow(<ReactComponent component={componentWithEverything} depth={3} />, {
			...options,
			context: {
				config: {
					showUsage: true,
				},
			},
		});

		expect(actual.prop('tabButtons').props.active).toBe(DOCS_TAB_USAGE);
		expect(actual.prop('tabBody').props.active).toBe(DOCS_TAB_USAGE);
	});

	it('should return null when component has no name', () => {
		const actual = shallow(
			<ReactComponent component={{ slug: 'foo', props: {} }} depth={3} />,
			options
		);

		expect(actual.getElement()).toBe(null);
	});

	test('should not render component in isolation mode by default', () => {
		const actual = shallow(<ReactComponent component={component} depth={3} />, options);

		expect(actual.prop('heading').props.slotProps.isolated).toBeFalsy();
	});

	test('should render component in isolation mode', () => {
		const actual = shallow(<ReactComponent component={component} depth={3} />, {
			context: {
				...options.context,
				displayMode: DisplayModes.component,
			},
		});

		expect(actual.prop('heading').props.slotProps.isolated).toBeTruthy();
	});

	it('should pass depth to heading', () => {
		const actual = shallow(<ReactComponent component={component} depth={3} />, options);

		expect(actual.prop('heading').props.depth).toBe(3);
	});

	it('should not render heading as deprecated by default', () => {
		const actual = shallow(<ReactComponent component={component} depth={3} />, options);

		expect(actual.prop('heading').props.deprecated).toBeFalsy();
	});

	it('should render heading as deprecated when @deprecated is present in tags', () => {
		const actual = shallow(
			<ReactComponent
				component={{
					...component,
					props: {
						tags: {
							deprecated: [
								{
									title: 'deprecated',
									description: 'I am deprecated',
								},
							],
						},
					},
				}}
				depth={3}
			/>,
			options
		);

		expect(actual.prop('heading').props.deprecated).toBeTruthy();
	});
});

describe('ReactComponentRenderer', () => {
	const props = {
		classes: {},
		name: 'Test',
		slug: 'test',
		pathLine: 'components/test',
		heading: <div>heading</div>,
	};

	test('should render component', () => {
		const actual = shallow(<ReactComponentRenderer {...props} />);

		expect(actual).toMatchSnapshot();
	});

	test('should render component without a pathline', () => {
		const actual = shallow(<ReactComponentRenderer {...props} pathLine="" />);

		expect(actual).toMatchSnapshot();
	});

	test('should render usage section', () => {
		const actual = shallow(
			<ReactComponentRenderer
				{...props}
				tabButtons={<div>tab buttons</div>}
				tabBody={<div>tab body</div>}
			/>
		);

		expect(actual).toMatchSnapshot();
	});

	test('should render description', () => {
		const actual = shallow(
			<ReactComponentRenderer {...props} description={<div>description</div>} />
		);

		expect(actual).toMatchSnapshot();
	});

	test('should render docs', () => {
		const actual = shallow(<ReactComponentRenderer {...props} docs={<div>docs</div>} />);

		expect(actual).toMatchSnapshot();
	});

	test('should render examples', () => {
		const actual = shallow(
			<ReactComponentRenderer
				{...props}
				examples={[<div key={1}>example 1</div>, <div key={2}>example 2</div>]}
			/>
		);

		expect(actual).toMatchSnapshot();
	});
});
