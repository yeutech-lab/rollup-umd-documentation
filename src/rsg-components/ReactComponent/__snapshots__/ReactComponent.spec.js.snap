// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactComponent should pass rendered description, usage, examples, etc. to the renderer 1`] = `
<Styled(ReactComponent)
  description={
    <Markdown
      text="Bar"
    />
  }
  docs={<JsDoc />}
  examples={
    <Examples
      examples={
        Array [
          Object {
            "content": 
<button>
  OK
</button>
,
            "evalInContext": [Function],
            "type": "code",
          },
          Object {
            "content": "Hello *world*!",
            "type": "markdown",
          },
        ]
      }
      name="Foo"
    />
  }
  heading={
    <SectionHeading
      deprecated={false}
      depth={3}
      id="foo"
      slotName="componentToolbar"
      slotProps={
        Object {
          "isolated": false,
          "metadata": Object {
            "tags": Array [
              "one",
              "two",
            ],
          },
          "name": "Foo",
          "pathLine": "foo/bar.js",
          "props": Object {
            "description": "Bar",
            "examples": Array [
              Object {
                "content": 
<button>
  OK
</button>
,
                "evalInContext": [Function],
                "type": "code",
              },
              Object {
                "content": "Hello *world*!",
                "type": "markdown",
              },
            ],
            "methods": Array [
              Object {
                "description": "Sets the counter to a particular value.",
                "name": "set",
                "params": Array [
                  Object {
                    "description": "New value for the counter.",
                    "name": "newValue",
                    "type": Object {
                      "name": "Number",
                    },
                  },
                ],
                "returns": null,
              },
            ],
          },
          "slug": "foo",
        }
      }
    >
      Foo
    </SectionHeading>
  }
  name="Foo"
  pathLine="foo/bar.js"
  slug="foo"
  tabBody={
    <Slot
      active={undefined}
      name="docsTabs"
      onlyActive={true}
      props={
        Object {
          "metadata": Object {
            "tags": Array [
              "one",
              "two",
            ],
          },
          "name": "Foo",
          "pathLine": "foo/bar.js",
          "props": Object {
            "description": "Bar",
            "examples": Array [
              Object {
                "content": 
<button>
  OK
</button>
,
                "evalInContext": [Function],
                "type": "code",
              },
              Object {
                "content": "Hello *world*!",
                "type": "markdown",
              },
            ],
            "methods": Array [
              Object {
                "description": "Sets the counter to a particular value.",
                "name": "set",
                "params": Array [
                  Object {
                    "description": "New value for the counter.",
                    "name": "newValue",
                    "type": Object {
                      "name": "Number",
                    },
                  },
                ],
                "returns": null,
              },
            ],
          },
          "slug": "foo",
        }
      }
    />
  }
  tabButtons={
    <Slot
      active={undefined}
      name="docsTabButtons"
      props={
        Object {
          "metadata": Object {
            "tags": Array [
              "one",
              "two",
            ],
          },
          "name": "Foo",
          "onClick": [Function],
          "pathLine": "foo/bar.js",
          "props": Object {
            "description": "Bar",
            "examples": Array [
              Object {
                "content": 
<button>
  OK
</button>
,
                "evalInContext": [Function],
                "type": "code",
              },
              Object {
                "content": "Hello *world*!",
                "type": "markdown",
              },
            ],
            "methods": Array [
              Object {
                "description": "Sets the counter to a particular value.",
                "name": "set",
                "params": Array [
                  Object {
                    "description": "New value for the counter.",
                    "name": "newValue",
                    "type": Object {
                      "name": "Number",
                    },
                  },
                ],
                "returns": null,
              },
            ],
          },
          "slug": "foo",
        }
      }
    />
  }
/>
`;

exports[`ReactComponentRenderer should render component 1`] = `
<div
  id="Test-container"
>
  <header>
    <div>
      heading
    </div>
    <Styled(Pathline)>
      components/test
    </Styled(Pathline)>
  </header>
</div>
`;

exports[`ReactComponentRenderer should render component without a pathline 1`] = `
<div
  id="Test-container"
>
  <header>
    <div>
      heading
    </div>
  </header>
</div>
`;

exports[`ReactComponentRenderer should render description 1`] = `
<div
  id="Test-container"
>
  <header>
    <div>
      heading
    </div>
    <Styled(Pathline)>
      components/test
    </Styled(Pathline)>
  </header>
  <div>
    <div>
      description
    </div>
  </div>
</div>
`;

exports[`ReactComponentRenderer should render docs 1`] = `
<div
  id="Test-container"
>
  <header>
    <div>
      heading
    </div>
    <Styled(Pathline)>
      components/test
    </Styled(Pathline)>
  </header>
  <div>
    <div>
      docs
    </div>
  </div>
</div>
`;

exports[`ReactComponentRenderer should render examples 1`] = `
<div
  id="Test-container"
>
  <header>
    <div>
      heading
    </div>
    <Styled(Pathline)>
      components/test
    </Styled(Pathline)>
  </header>
  <div
    key="1"
  >
    example 1
  </div>
  <div
    key="2"
  >
    example 2
  </div>
</div>
`;

exports[`ReactComponentRenderer should render usage section 1`] = `
<div
  id="Test-container"
>
  <header>
    <div>
      heading
    </div>
    <Styled(Pathline)>
      components/test
    </Styled(Pathline)>
  </header>
  <div>
    <div>
      <div>
        tab buttons
      </div>
    </div>
    <div>
      tab body
    </div>
  </div>
</div>
`;
